### PR TITLE
JSON API for some endpoints

### PR DESCRIPTION
**Pull request by PizieDust**
This PR attempts to add endpoints that return json. 
- `/job/:job/build/latest-uuid` -> returns the uuid of the latest jon
- `/search-uuid` -> returns the uuid from a sha256
- `/compare/json/:build_left/:build_right` -> returns the result of a comparison in json

This is related to https://github.com/robur-coop/mollymawk/issues/54

### CODE REVIEW

File: lib/builder_web.ml [hunk 1]
```diff
@@ -352,6 +352,39 @@ let routes ~datadir ~cachedir ~configdir ~expired_jobs =
     let job_name = Dream.param req "job" in
     let platform = Dream.query req "platform" in
     redirect_latest req ~job_name ~platform ~artifact:""
+
```
Comment (hannesm) [inline lib/builder_web.ml:355]:
```suggestion
```

spurious newline

File: lib/builder_web.ml [hunk 2]
```diff
@@ -352,6 +352,39 @@ let routes ~datadir ~cachedir ~configdir ~expired_jobs =
...
+  in
+
+  let return_latest_uuid req =
+    let job_name = Dream.param req "job" in
+    let platform = Dream.query req "platform" in
+    (Dream.sql req (Model.job_id job_name) >>= Model.not_found >>= fun job_id ->
+      Dream.sql req (Model.latest_successful_build_uuid job_id platform))
+      >>= Model.not_found
+      |> if_error "Error getting job" >>= fun build ->
```
Comment (hannesm) [inline lib/builder_web.ml:364]: should the error case as return a json object?

File: lib/builder_web.ml [hunk 3]
```diff
@@ -352,6 +352,39 @@ let routes ~datadir ~cachedir ~configdir ~expired_jobs =
...
+    begin try Ohex.decode hash_hex |> Lwt_result.return
+      with Invalid_argument e -> Lwt_result.fail (`Msg ("Bad hex: " ^ e))
+    end
+    |> if_error ~status:`Bad_Request "Bad request" >>= fun hash ->
+    Dream.sql req (Model.build_hash hash) >>= Model.not_found
+    |> if_error "Internal server error" >>= fun (job_name, build) ->
+      let json_response =
+        `Assoc [
+          "uuid", `String (Uuidm.to_string (build.Builder_db.Build.uuid));
+          "job", `String job_name
```
Comment (reynir) [inline lib/builder_web.ml:384]: Would it make sense to return more information from the build? For example, we could return the platform and execution {start,finish}, and maybe even if there's a main binary. For more information (e.g. other build artifacts) we would need to do more database queries...

File: lib/builder_web.ml [hunk 4]
```diff
@@ -352,6 +352,39 @@ let routes ~datadir ~cachedir ~configdir ~expired_jobs =
     let job_name = Dream.param req "job" in
     let platform = Dream.query req "platform" in
     redirect_latest req ~job_name ~platform ~artifact:""
+
+  in
+
+  let return_latest_uuid req =
```
Comment (hannesm) [inline lib/builder_web.ml:358]:
```suggestion
  let latest_uuid req =
```

I'd not prefix the function name with "return".

File: lib/builder_web.ml [hunk 5]
```diff
@@ -352,6 +352,39 @@ let routes ~datadir ~cachedir ~configdir ~expired_jobs =
...
+      Dream.sql req (Model.latest_successful_build_uuid job_id platform))
+      >>= Model.not_found
+      |> if_error "Error getting job" >>= fun build ->
+        let json_response =
+          `Assoc [ "uuid", `String (Uuidm.to_string build) ] |> Yojson.Basic.to_string
+        in
+        Dream.json ~status:`OK json_response |> Lwt_result.ok
+  in
+
+  let return_uuid req =
```
Comment (hannesm) [inline lib/builder_web.ml:371]:
```suggestion
  let search_uuid req =
```

File: lib/builder_web.ml [hunk 6]
```diff
@@ -352,6 +352,39 @@ let routes ~datadir ~cachedir ~configdir ~expired_jobs =
...
     let platform = Dream.query req "platform" in
     redirect_latest req ~job_name ~platform ~artifact:""
+
+  in
+
+  let return_latest_uuid req =
+    let job_name = Dream.param req "job" in
+    let platform = Dream.query req "platform" in
+    (Dream.sql req (Model.job_id job_name) >>= Model.not_found >>= fun job_id ->
+      Dream.sql req (Model.latest_successful_build_uuid job_id platform))
```
Comment (hannesm) [inline lib/builder_web.ml:362]: there's some code duplication now -- should that be shared (between redirect_latest and this function here).

File: lib/builder_web.ml [hunk 7]
```diff
@@ -352,6 +352,39 @@ let routes ~datadir ~cachedir ~configdir ~expired_jobs =
...
+    |> if_error ~status:`Bad_Request "Bad request" >>= fun hash ->
+    Dream.sql req (Model.build_hash hash) >>= Model.not_found
+    |> if_error "Internal server error" >>= fun (job_name, build) ->
+      let json_response =
+        `Assoc [
+          "uuid", `String (Uuidm.to_string (build.Builder_db.Build.uuid));
+          "job", `String job_name
+        ] |> Yojson.Basic.to_string
+      in
+      Dream.json ~status:`OK json_response |> Lwt_result.ok
```
Comment (hannesm) [inline lib/builder_web.ml:387]: same comments apply: should the error cases return json as well? should the code be shared?

File: lib/builder_web.ml [hunk 8]
```diff
@@ -572,6 +605,13 @@ let routes ~datadir ~cachedir ~configdir ~expired_jobs =
     let switch_left = OpamFile.SwitchExport.read_from_string switch_left
     and switch_right = OpamFile.SwitchExport.read_from_string switch_right in
     let opam_diff = Opamdiff.compare switch_left switch_right in
+    (job_left, job_right, build_left, build_right, env_diff, pkg_diff, opam_diff) |> Lwt.return_ok
+
```
Comment (hannesm) [inline lib/builder_web.ml:609]:
```suggestion
```

spurious newline

File: lib/builder_web.ml [hunk 9]
```diff
@@ -627,6 +694,8 @@ let routes ~datadir ~cachedir ~configdir ~expired_jobs =
     `Get, "/job/:job/failed", (w job_with_failed);
     `Get, "/job/:job/build/latest/**", (w redirect_latest);
     `Get, "/job/:job/build/latest", (w redirect_latest_no_slash);
+    `Get, "/job/:job/build/latest-uuid", (w return_latest_uuid);
+    `Get, "/search-uuid", (w return_uuid);
```
Comment (reynir) [inline lib/builder_web.ml:698]: I think I'm maybe more inclined to have "/job/:job/build/:build" return a generic build info object with uuid, (job,) platform when there's an `Accept: application/json` or so. And then you would use the existing "/hash" and "/job/:job/build/latest" endpoints and follow the redirects.

File: lib/builder_web.ml [hunk 10]
```diff
@@ -639,6 +708,7 @@ let routes ~datadir ~cachedir ~configdir ~expired_jobs =
     `Get, "/all-builds", (w (builds ~all:true));
     `Get, "/hash", (w hash);
     `Get, "/compare/:build_left/:build_right", (w compare_builds);
+    `Get, "/compare/json/:build_left/:build_right", (w compare_builds_json);
```
Comment (hannesm) [inline lib/builder_web.ml:711]: the alternative, that Reynir suggested, was to use the same endpoint, and respect the `Accept` header (whether it contains `application/json` or not.

Comment (reynir) [inline lib/builder_web.ml:711]: Here as well we can check for `Accept:` header and serve json or html and keep it as one endpoint.

File: opamdiff/opamdiff.mli [hunk 11]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (hannesm) [inline opamdiff/opamdiff.mli:1]:
```suggestion
```

File: opamdiff/opamdiff.mli [hunk 12]
```diff
@@ -37,3 +38,6 @@ val compare: OpamFile.SwitchExport.t ->
   OpamFile.SwitchExport.t ->
   opam_diff list * version_diff list * OpamPackage.Set.t * OpamPackage.Set.t * ((string * string) list * (string * string) list * duniverse_diff list, [> `Msg of string ]) result
 
+val compare_result_to_json: 'a list * version_diff list * OpamPackage.Set.t * OpamPackage.Set.t *
+(('b * 'b) list * ('b * 'b) list * 'c list, [< `Msg of 'd ]) result -> Yojson.Basic.t
+
```
Comment (hannesm) [inline opamdiff/opamdiff.mli:43]:
```suggestion
```

**Review by reynir — COMMENTED** [review]
I think an API would be really nice, so thank you for taking first steps here!

I am thinking it could be nicer to reuse some of the existing endpoints and check the `Accept:` header instead. This duplicates code a bit.

Finally, I know sometimes people version their APIs by the path, e.g. `/api/v1/...` and I wonder if we should do the same. After all, this API is likely to be consumed by at least both mollymawk and albatross. For me it would be acceptable to go with an initial unversioned API and then do a breaking change if we don't want to make decisions now.

**Pull request closed without merging**